### PR TITLE
[bazel] Add a config-driven splice rule

### PR DESCRIPTION
**Pull request by cfrantz**
1. Create a localtools toolchain that transitions to a "golden" host
   configuration.  This is done so that we have exactly one
   configuration for tools and we wont have to worry about build flags
   propagating into `cfg=exec` dependencies.
2. Write a universal splice rule.
3. Add a config transition that sets the ROM, OTP and exec_env to
   splice.
4. Hide the splicing parameters from the `host` and `opentitan` plaform
   transitions to prevent the proliferation of build artifacts per build
   flags.
5. Update the exec_env configurations to supply base bitstreams,
   MMI files and ROM/OTP configurations.
6. Update the cw310_params block to always supply the universal
   splice rule if no other bitstream is requested.
7. Eliminate explicit splices in favor of implicit splices provided by the universal splice rule.

### CODE REVIEW

File: hw/top_earlgrey/BUILD [hunk 1]
```diff
@@ -85,26 +89,28 @@ fpga_cw310(
     name = "fpga_cw310_rom_with_fake_keys",
     testonly = True,
     base = ":fpga_cw310_test_rom",
-    bitstream = "//hw/bitstream:rom_with_fake_keys",
+    #otp = "//hw/ip/otp_ctrl/data:img_rma",
```
Comment (timothytrippel) [inline hw/top_earlgrey/BUILD:92]: comment?

File: sw/device/silicon_creator/rom_ext/e2e/verified_boot/BUILD [hunk 2]
```diff
@@ -92,7 +91,7 @@ _KEYS = {
         name = "key_{}".format(name),
         srcs = ["boot_test.c"],
         cw310 = cw310_params(
-            bitstream = "//sw/device/silicon_creator/rom_ext/e2e:bitstream_secret2_locked",
+            #bitstream = "//sw/device/silicon_creator/rom_ext/e2e:bitstream_secret2_locked",
```
Comment (timothytrippel) [inline sw/device/silicon_creator/rom_ext/e2e/verified_boot/BUILD:94]: comment?

File: sw/device/silicon_creator/rom/e2e/jtag_inject/BUILD [hunk 3]
```diff
@@ -120,13 +100,13 @@ LC_STATES_DEBUG_DISALLOWED = [
...
         cw310 = cw310_jtag_params(
             timeout = "moderate",
-            binaries = {
-                "//sw/device/silicon_creator/rom:rom_with_fake_keys_fpga_cw310_elf": "rom",
-            },
-            bitstream = ":rom_with_fake_keys_otp_{}_exec_disabled".format(lc_state),
             logging = "debug",
+            #binaries = {
+            #    "//sw/device/silicon_creator/rom:rom_with_fake_keys_fpga_cw310_elf": "rom",
+            #},
```
Comment (timothytrippel) [inline sw/device/silicon_creator/rom/e2e/jtag_inject/BUILD:106]: comment?

**Review by timothytrippel — APPROVED** [review]

**Review by moidx — APPROVED** [review]
This is a great improvement @cfrantz! It is also going to help us make the FPGA SiVal test targets easier to compose based on OTP dependencies. Thanks!

**Review by pamaury — APPROVED** [review]

File: hw/top_earlgrey/BUILD (hunk 1 above)
Comment (cfrantz) [reply hw/top_earlgrey/BUILD:92]: Removed

File: sw/device/silicon_creator/rom/e2e/jtag_inject/BUILD (hunk 3 above)
Comment (cfrantz) [reply sw/device/silicon_creator/rom/e2e/jtag_inject/BUILD:106]: Removed

File: rules/opentitan/fpga_cw310.bzl [hunk 4]
```diff
@@ -226,6 +230,10 @@ def cw310_params(
     Returns:
       struct of test parameters.
     """
+    if bitstream and (rom or otp):
+        fail("Cannot use rom or otp with bitstream.")
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply rules/opentitan/fpga_cw310.bzl:234]: The baseline bitstream is handled by the `base_bitstream` parameter in the exec_env.  The exec_env supplies the base bitstream and the required mmi files, thus allowing an appropriate splice for each fpga target.

The point of leaving `bitstream` as a parameter in the `cw310_params` (or any params) block is to allow a full bitstream override if that is what is required by the test.

File: rules/opentitan/splice.bzl [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+# Rules for memory splicing with Vivado.
+
+DEFAULTS = struct(
+    rom = "//hw/bitstream/universal:none",
+    otp = "//hw/bitstream/universal:none",
+    env = "//hw/bitstream/universal:none",
+)
+
+def gen_vivado_image(ctx, name, src, tool, swap_nibbles = True):
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply rules/opentitan/splice.bzl:19]: Renamed

File: rules/opentitan/splice.bzl [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        use_default_shell_env = False,
+        execution_requirements = {
+            "no-sandbox": "",
+        },
+        env = ENV,
+    )
+    return spliced
+
+def update_usr_access(ctx, name, src, tool):
+    output = ctx.actions.declare_file("{}.bit".format(name))
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply rules/opentitan/splice.bzl:73]: done

**Pull request merged**